### PR TITLE
Support sequential focus navigation

### PR DESCRIPTION
**Pull request by nerith**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13460)

### CODE REVIEW

**Comment by jdm** [comment]
I'm really sorry I haven't reviewed this yet. In the interests of allowing it to make progress while I unearth myself from lots of review requests, I'm handing it off.
r? @nox

File: components/script/dom/document.rs [hunk 1]
```diff
@@ -1656,6 +1668,64 @@ impl Document {
...
+        // TODO: Step 4
+
+        // Step 5
+        let candidate = self.sequential_search(direction);
+
+        // Step 6
+        self.request_focus(&self.focus_list.borrow()[candidate]);
+    }
+
+    fn sequential_search(&self, direction: NavigationDirection) -> usize {
```
Comment (nox) [inline components/script/dom/document.rs]: This is supposed to return a candidate; shouldn't the return type be `Root<Element>`?

Comment (nox) [inline components/script/dom/document.rs]: To which behaviour this method correspond in the spec for that algorithm?

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -1975,6 +2047,30 @@ impl Document {
...
 }
 
+impl VirtualMethods for Document {
+    fn super_type(&self) -> Option<&VirtualMethods> {
+        Some(self.upcast::<Document>() as &VirtualMethods)
+    }
+
+    fn handle_event(&self, event: &Event) {
+        if event.type_() == atom!("keydown") {
+            let key_event = event.downcast::<KeyboardEvent>().unwrap();
```
Comment (nox) [inline components/script/dom/document.rs]: Try to downcast to `KeyboardEvent`, and only then check that its type is `keydown`.

File: components/script/dom/element.rs [hunk 3]
```diff
@@ -2203,6 +2203,19 @@ impl VirtualMethods for Element {
...
+            let children = self.upcast::<Node>().traverse_preorder();
+
+            for child in children {
+                if let Some(element) = child.downcast::<Element>() {
+                    if element.is_focusable_area() && !element.disabled_state() {
+                        doc.add_focusable_element(element);
+                    }
+                }
+            }
+        }
```
Comment (nox) [inline components/script/dom/element.rs]: When do you remove focusable elements from the document? Shouldn't that be done in `unbind_from_tree`?

File: components/script/dom/htmlanchorelement.rs [hunk 4]
```diff
@@ -99,6 +100,20 @@ impl VirtualMethods for HTMLAnchorElement {
...
+                event.downcast::<KeyboardEvent>() {
+                    match keyevent.get_key() {
+                        Some(Key::Enter) => {
+                            follow_hyperlink(self.upcast::<Element>(), Some(String::new()), None);
+                        },
+                        _ => ()
+                    }
+            }
+        }
+    }
```
Comment (nox) [inline components/script/dom/htmlanchorelement.rs]: Could that change come with a spec link?

File: components/script/dom/htmlanchorelement.rs [hunk 5]
```diff
@@ -99,6 +100,20 @@ impl VirtualMethods for HTMLAnchorElement {
             _ => self.super_type().unwrap().parse_plain_attribute(name, value),
         }
     }
+
+    fn handle_event(&self, event: &Event) {
```
Comment (nox) [inline components/script/dom/htmlanchorelement.rs]: You didn't call the super type `handle_event` method.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by nerith** [comment]
@KiChjang I'm still working on this. Sorry for the delay. I had hardware issues the past week and was blocked on my machine by issue #13823.

**Comment by nerith** [comment]
Review status: 0 of 3 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[components/script/dom/document.rs, line 244 at r3](https://reviewable.io:443/reviews/servo/servo/13460#-KSg00VQMEXjXTVsWI5Q-r3-244:-KUZc0d_nXq0qgm1Hszi:but0nel) ([raw file](https://github.com/servo/servo/blob/acdbb2ee334f7999be71f7c624cb71309a9889fa/components/script/dom/document.rs#L244)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Why `u64` and not `usize`?
</details>
`usize` will work. I don't remember why I made this `u64`.

---

*[components/script/dom/document.rs, line 1703 at r3](https://reviewable.io:443/reviews/servo/servo/13460#-KSg00VQMEXjXTVsWI5Q-r3-1703:-KUZcaychXqLM7jFIFmn:b-8jea7r) ([raw file](https://github.com/servo/servo/blob/acdbb2ee334f7999be71f7c624cb71309a9889fa/components/script/dom/document.rs#L1703)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> To which behaviour this method correspond in the spec for that algorithm?
</details>
This is fixed now. I wasn't sure how to return a `Root<Element>` without getting a compiler error in the original changes. That's why I just returned the index.

Currently, the method is only trying to handle the "sequential" behavior specified in the sequential navigation search algorithm spec. I still need to specify `starting point` and `selection mechanism`.

---

*[components/script/dom/element.rs, line 2217 at r3](https://reviewable.io:443/reviews/servo/servo/13460#-KSg00VRfvdMxWwdx2Ui-r3-2217:-KWi8uxWRbkwCq1HwKYX:b-rzgiw8) ([raw file](https://github.com/servo/servo/blob/acdbb2ee334f7999be71f7c624cb71309a9889fa/components/script/dom/element.rs#L2217)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> When do you remove focusable elements from the document? Shouldn't that be done in `unbind_from_tree`?
</details>
I did not handle removing elements from the ordering list.  I added a function called `remove_focusable_element` in document.rs and call it in `unbind_from_tree` now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13460)*

**Comment by jdm** [comment]
@nox Review ping.

**Comment by jdm** [comment]
@nox Post holiday review ping.

File: components/script/dom/htmlanchorelement.rs [hunk 6]
```diff
@@ -559,7 +577,7 @@ fn is_current_browsing_context(target: DOMString) -> bool {
 }
 
 /// https://html.spec.whatwg.org/multipage/#following-hyperlinks-2
-fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>, referrer_policy: Option<ReferrerPolicy>) {
+pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>, referrer_policy: Option<ReferrerPolicy>) {
```
Comment (nox) [inline components/script/dom/htmlanchorelement.rs]: AFAIK this doesn't need to be public.

**Comment by nox** [comment]
I'll carry this in the next few days.

**Comment by nox** [comment]
Superseded by #15936, thanks for your contribution!

**Pull request closed without merging**